### ISSUE TITLE
How to skip one stage from multi-stage docker build

### ISSUE THREAD

**Issue by venkat-koppisetti** [docker/cli#1134]
**Description**

We have Multistage docker build that creates rpm in each stages. if one rpm is there in folder i dont want to build that particular stage.
So how to skip one stage from multistage docker
is there any condition in dockerfile like if, to skip one stage or few docker commands 

**Describe the results you expected:**

Skip one stage from multi-stage docker build

**Output of `docker version`:**

[root@localhost consul]# docker -v
Docker version 18.03.0-ce, build 0520e24

**Output of `docker info`:**

Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 105
Server Version: 18.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfd04396dc68220d1cecbe686a6cc3aa5ce3667c
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859GiB
Name: localhost.localdomain
ID: QSFU:ODKS:LJGZ:GC34:KXTP:6B7Y:5UMB:Q7WT:V2X3:4K6M:DFLQ:I7WS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running docker in Centos7 VirtualBox

**Comment by cpuguy83** [comment]
There is no way to skip stages per se.
You can build specific stages: `docker build --target=<stage>`.

Why are you creating the same thing in multiple stages? You can have a stage just for creating that thing and then copy it into the stage that requires it.

**Comment by venkat-koppisetti** [comment]
Each stage has performs different build. Suppose in one stage we are creating rpm, we dont wanna recreate rpm using that build if rpm is already there, in that case we want to skip that stage

**Comment by thaJeztah** [comment]
> we dont wanna recreate rpm using that build if rpm is already there, in that case we want to skip that stage .

Not sure I understand the "if rpm is already there" part. Docker build will cache Dockerfile steps where possible.

BuildKit (which will be included as an experimental feature in the next docker release) will further optimize that, and skip stages that are not needed for the `--target` stage

**Comment by marcellodesales** [comment]
* @venkat-koppisetti , @cpuguy83 
* We have the similar need to skip stages
* All our Dockerfiles are also multi-stage and we build using Jenkinsfile

# Requirement

* Each stage builds images to run `unit` and `integration` test images
* Each stage is run in parallel in a Jenkins pipeline
* When all the different test containers pass, we need to build an image with the binary (war, react, etc)
** The stage image that is meant for publishing a binary is still needed to cover extra requirements such as reports, older monolith support, etc
** Skipping the stages not part of the build is important. We want to avoid `docker build` from running a specific stage that's not meant to be run.
** We'd like to specify in Dockerfile that the stage is optional

**Comment by mauricios** [comment]
I have the following Dockerfile with multiple stages. For efficiency I'm running different tasks in different stages so I want to skip some stages depending on the needs of the final stage.

```dockerfile
# Defining environment
ARG APP_ENV=dev

# Building the base image
FROM alpine as base
RUN echo "running BASE commands"

# Building de pre-install Prod image
FROM base as prod-preinstall
RUN echo "running PROD pre-install commands"

# Building the Dev image
FROM base as dev-preinstall
RUN echo "running DEV pre-install commands"

# Installing the app files
FROM ${APP_ENV}-preinstall as install
COPY app ./app
RUN echo "running install commands"

FROM install as prod-postinstall
RUN echo "running PROD post-install commands"

FROM install as dev-postinstall
RUN echo "running DEV post-install commands"

FROM ${APP_ENV}-postinstall as final
RUN echo "running final commands"
```

If `APP_ENV=PROD` I want to skip all the stages that not depend on the PROD stages, not only for efficiency but because the `DEV` stages will break at some point due to missing stages. Now I know that skipping stages in not currently supported in Dockerfiles, and this scenario can be achieved using multiple Dockerfiles (with duplicated code) or adding some wrapper scripts to check the `APP_ENV` argument. 

This can be a pretty good feature to build more complex and efficient Docker images.

**Comment by thaJeztah** [comment]
@mauricios yes; BuildKit does exactly that; it will only build stages that are needed to build the final stage (or stage that's specified through `--target`). BuildKit is integrated in Docker 18.06 as an experimental feature (release candidates for 18.06 are already available; Docker for Mac/Windows releases will be available soon).

**Comment by thaJeztah** [comment]
Docker 18.06 has been released, and includes experimental BuildKit support; to use this;

- enable experimental features on the daemon (`{"experimental":true})` in the `daemon.json` configuration file
- set `DOCKER_BUILDKIT=1` environment variable to use buildkit (remove the environment variable, or set it to `0` to disable)

Here's what it looks like;

with `APP_ENV=prod`:

```bash
DOCKER_BUILDKIT=1 docker build -t prod --build-arg APP_ENV=prod -<<'EOF'
# Defining environment
ARG APP_ENV=dev

# Building the base image
FROM alpine as base
RUN echo "running BASE commands"

# Building de pre-install Prod image
FROM base as prod-preinstall
RUN echo "running PROD pre-install commands"

# Building the Dev image
FROM base as dev-preinstall
RUN echo "running DEV pre-install commands"

# Installing the app files
FROM ${APP_ENV}-preinstall as install
RUN echo "running install commands"

FROM install as prod-postinstall
RUN echo "running PROD post-install commands"

FROM install as dev-postinstall
RUN echo "running DEV post-install commands"

FROM ${APP_ENV}-postinstall as final
RUN echo "running final commands"
EOF
```

Only the PROD build-stages are executed:

```
[+] Building 2.1s (9/9) FINISHED                                       
 => local://context (.dockerignore)                               0.0s
 => => transferring context: 02B                                  0.0s
 => local://dockerfile (Dockerfile)                               0.0s
 => => transferring dockerfile: 705B                              0.0s
 => docker-image://docker.io/library/alpine:latest                0.0s
 => => resolve docker.io/library/alpine:latest                    0.0s
 => /bin/sh -c echo "running BASE commands"                       0.5s
 => /bin/sh -c echo "running PROD pre-install commands"           0.4s
 => /bin/sh -c echo "running install commands"                    0.4s
 => /bin/sh -c echo "running PROD post-install commands"          0.4s
 => /bin/sh -c echo "running final commands"                      0.4s
 => exporting to image                                            0.0s
 => => exporting layers                                           0.0s
 => => writing image sha256:57048e9c9845084931f74490091563aadc50  0.0s
 => => naming to docker.io/library/foo                            0.0s

```

with `APP_ENV=dev`:


```bash
DOCKER_BUILDKIT=1 docker build -t dev -<<'EOF'
# Defining environment
ARG APP_ENV=dev

# Building the base image
FROM alpine as base
RUN echo "running BASE commands"

# Building de pre-install Prod image
FROM base as prod-preinstall
RUN echo "running PROD pre-install commands"

# Building the Dev image
FROM base as dev-preinstall
RUN echo "running DEV pre-install commands"

# Installing the app files
FROM ${APP_ENV}-preinstall as install
RUN echo "running install commands"

FROM install as prod-postinstall
RUN echo "running PROD post-install commands"

FROM install as dev-postinstall
RUN echo "running DEV post-install commands"

FROM ${APP_ENV}-postinstall as final
RUN echo "running final commands"
EOF
```

Only the DEV build-stages are executed:

```
[+] Building 1.8s (9/9) FINISHED                                         
 => local://dockerfile (Dockerfile)                                 0.0s
 => => transferring dockerfile: 705B                                0.0s
 => local://context (.dockerignore)                                 0.0s
 => => transferring context: 02B                                    0.0s
 => docker-image://docker.io/library/alpine:latest                  0.0s
 => CACHED /bin/sh -c echo "running BASE commands"                  0.0s
 => /bin/sh -c echo "running DEV pre-install commands"              0.4s
 => /bin/sh -c echo "running install commands"                      0.5s
 => /bin/sh -c echo "running DEV post-install commands"             0.4s
 => /bin/sh -c echo "running final commands"                        0.4s
 => exporting to image                                              0.0s
 => => exporting layers                                             0.0s
 => => writing image sha256:1cf6caaaab4689f6b11ea4d61b5b6175e54074  0.0s
 => => naming to docker.io/library/dev                              0.0s
```

**Comment by mhumeSF** [comment]
Very excited about this feature, but I am unable to get this running. Also, how do you get such tidy output? Running on docker-for-mac with 18.06-ce edge. `Version 18.06.0-ce-mac69 (26407)`

```
$ DOCKER_BUILDKIT=1 docker build -t prod --build-arg APP_ENV=prod -<<'EOF'
# Defining environment
ARG APP_ENV=dev

# Building the base image
FROM alpine as base
RUN echo "running BASE commands"

# Building de pre-install Prod image
FROM base as prod-preinstall
RUN echo "running PROD pre-install commands"

# Building the Dev image
FROM base as dev-preinstall
RUN echo "running DEV pre-install commands"

# Installing the app files
FROM ${APP_ENV}-preinstall as install
RUN echo "running install commands"

FROM install as prod-postinstall
RUN echo "running PROD post-install commands"

FROM install as dev-postinstall
RUN echo "running DEV post-install commands"

FROM ${APP_ENV}-postinstall as final
RUN echo "running final commands"
EOF
Sending build context to Docker daemon   2.56kB
Step 1/15 : ARG APP_ENV=dev
Step 2/15 : FROM alpine as base
latest: Pulling from library/alpine
8e3ba11ec2a2: Pull complete 
Digest: sha256:7043076348bf5040220df6ad703798fd8593a0918d06d3ce30c6c93be117e430
Status: Downloaded newer image for alpine:latest
 ---> 11cd0b38bc3c
Step 3/15 : RUN echo "running BASE commands"
 ---> Running in c1ce79f04266
running BASE commands
Removing intermediate container c1ce79f04266
 ---> 00dc93d70418
Step 4/15 : FROM base as prod-preinstall
 ---> 00dc93d70418
Step 5/15 : RUN echo "running PROD pre-install commands"
 ---> Running in a5b2261eb1a4
running PROD pre-install commands
Removing intermediate container a5b2261eb1a4
 ---> b69aac8a0873
Step 6/15 : FROM base as dev-preinstall
 ---> 00dc93d70418
Step 7/15 : RUN echo "running DEV pre-install commands"
 ---> Running in 9981b5cda531
running DEV pre-install commands
Removing intermediate container 9981b5cda531
 ---> 396748cfd570
Step 8/15 : FROM ${APP_ENV}-preinstall as install
 ---> b69aac8a0873
Step 9/15 : RUN echo "running install commands"
 ---> Running in 4f79ecebd995
running install commands
Removing intermediate container 4f79ecebd995
 ---> 0912780ecc31
Step 10/15 : FROM install as prod-postinstall
 ---> 0912780ecc31
Step 11/15 : RUN echo "running PROD post-install commands"
 ---> Running in 69bebc0522f8
running PROD post-install commands
Removing intermediate container 69bebc0522f8
 ---> 531c4c0342c7
Step 12/15 : FROM install as dev-postinstall
 ---> 0912780ecc31
Step 13/15 : RUN echo "running DEV post-install commands"
 ---> Running in 21319511956b
running DEV post-install commands
Removing intermediate container 21319511956b
 ---> abafc0717bec
Step 14/15 : FROM ${APP_ENV}-postinstall as final
 ---> 531c4c0342c7
Step 15/15 : RUN echo "running final commands"
 ---> Running in 26cdd387d0f7
running final commands
Removing intermediate container 26cdd387d0f7
 ---> 4c57a0819621
Successfully built 4c57a0819621
Successfully tagged prod:latest
```

**Comment by bwowk** [comment]
Being able to interpolate vars on the build directives is great, but maybe we could have something more seamless, like building only the depended-upon stages for each target.

The dependency could be inferred both from `COPY --from=<stage>` directives and `FROM <stage>` directives. So for example, with this Dockerfile:
```
# Compile module Foo
FROM gcc as foo
RUN compile-foo.sh

# Install runtime dependencies
FROM alpine as base
RUN apk install <dependencies> 

# Building img flavor A
FROM base as versionA
RUN [...]

# Building img flavor B using module Foo
FROM versionA as versionB
COPY --from=foo /lib/foo /lib/foo
```

We could build the dependency graph based on the target:

```
versionB --> versionA --> base
        `--> foo
```

So building with `versionA` as target would trigger the `base` and `versionA` stages.
Likewise, building with `versionB` as target would also trigger the `base` and `versionA` stages, but additionally it would trigger the `foo` stage, as the dependency is implied by the `COPY` directive.

**Comment by cpuguy83** [comment]
@bwowk this is exactly what buildkit does.

**Comment by bwowk** [comment]
Oh, my bad, I completely misinterpreted the example.

I guess @thaJeztah meant to write `ARG APP_ENV=prod` in the first sample Dockerfile: 
> with `APP_ENV=prod`:
> 
> ```shell
> DOCKER_BUILDKIT=1 docker build -t prod --build-arg APP_ENV=prod -<<'EOF'
> # Defining environment
> ARG APP_ENV=dev
> ```

That's great :)
Can't wait to see it included as a stable feature.

**Comment by thaJeztah** [comment]
> I guess @thaJeztah meant to write ARG APP_ENV=prod in the first sample Dockerfile:

No, that's not needed; the `ARG APP_ENV=dev` line sets the _default_ value for `APP_ENV`. That means that someone building the dockerfile, without setting a `--build-arg` will build a dev-image by default. To _override_ the default, provide a `--build-arg` flag.

**Comment by romafederico** [comment]
Hey @thaJeztah ... Thanks for that example. Quick question: Is it possible that you mean `--target` instead of `-t` in the following command?
```
DOCKER_BUILDKIT=1 docker build -t prod --build-arg APP_ENV=prod -<<'EOF'
```

I thought `-t` was for `--tag`...

**Comment by thaJeztah** [comment]
That's correct; `-t` is shorthand for `--tag`. In my example, both the `prod` and `dev` image build the `final` stage (target), but the `--build-arg` sets the value for `APP_ENV`. Therefore the `final` stage will use a different stage as base image

**Comment by keithmattix** [comment]
Did this ever get made a stable feature? I noticed the issue is still open

**Comment by thaJeztah** [comment]
Buildkit is available on current versions of docker (but does still require you to enable it with with the `DOCKER_BUILDKIT=1` environment variable)

**Comment by thaJeztah** [comment]
Let me close this issue, because this is addressed in buildkit, and not something that can be addressed in the classic builder

**Comment by njleonzhang** [comment]
@ thaJeztah Wait this feature to be released formally, it's a really helpful.

**Comment by njleonzhang** [comment]
@ thaJeztah Wait this feature to be released formally, it's a really helpful.

**Comment by thaJeztah** [comment]
@njleonzhang the feature has been released; BuildKit is included in Docker 18.09 and up (including the current 19.03 release); it's not enabled by default (see https://github.com/moby/moby/issues/40379), but it's ready to be used for building images (and generally recommended to use instead of the classic builder); enable buildkit by setting `DOCKER_BUILDKIT=1` in your environment

**Comment by NicholasFlamy** [comment]
> Docker 18.06 has been released, and includes experimental BuildKit support; to use this;
> 
> * enable experimental features on the daemon (`{"experimental":true})` in the `daemon.json` configuration file
> * set `DOCKER_BUILDKIT=1` environment variable to use buildkit (remove the environment variable, or set it to `0` to disable)
> 
> Here's what it looks like;
> 
> with `APP_ENV=prod`:
> 
> DOCKER_BUILDKIT=1 docker build -t prod --build-arg APP_ENV=prod -<<'EOF'
> # Defining environment
> ARG APP_ENV=dev
> 
> # Building the base image
> FROM alpine as base
> RUN echo "running BASE commands"
> 
> # Building de pre-install Prod image
> FROM base as prod-preinstall
> RUN echo "running PROD pre-install commands"
> 
> # Building the Dev image
> FROM base as dev-preinstall
> RUN echo "running DEV pre-install commands"
> 
> # Installing the app files
> FROM ${APP_ENV}-preinstall as install
> RUN echo "running install commands"
> 
> FROM install as prod-postinstall
> RUN echo "running PROD post-install commands"
> 
> FROM install as dev-postinstall
> RUN echo "running DEV post-install commands"
> 
> FROM ${APP_ENV}-postinstall as final
> RUN echo "running final commands"
> EOF
> Only the PROD build-stages are executed:
> 
> ```
> [+] Building 2.1s (9/9) FINISHED                                       
>  => local://context (.dockerignore)                               0.0s
>  => => transferring context: 02B                                  0.0s
>  => local://dockerfile (Dockerfile)                               0.0s
>  => => transferring dockerfile: 705B                              0.0s
>  => docker-image://docker.io/library/alpine:latest                0.0s
>  => => resolve docker.io/library/alpine:latest                    0.0s
>  => /bin/sh -c echo "running BASE commands"                       0.5s
>  => /bin/sh -c echo "running PROD pre-install commands"           0.4s
>  => /bin/sh -c echo "running install commands"                    0.4s
>  => /bin/sh -c echo "running PROD post-install commands"          0.4s
>  => /bin/sh -c echo "running final commands"                      0.4s
>  => exporting to image                                            0.0s
>  => => exporting layers                                           0.0s
>  => => writing image sha256:57048e9c9845084931f74490091563aadc50  0.0s
>  => => naming to docker.io/library/prod                            0.0s
> ```
> 
> with `APP_ENV=dev`:
> 
> DOCKER_BUILDKIT=1 docker build -t dev -<<'EOF'
> # Defining environment
> ARG APP_ENV=dev
> 
> # Building the base image
> FROM alpine as base
> RUN echo "running BASE commands"
> 
> # Building de pre-install Prod image
> FROM base as prod-preinstall
> RUN echo "running PROD pre-install commands"
> 
> # Building the Dev image
> FROM base as dev-preinstall
> RUN echo "running DEV pre-install commands"
> 
> # Installing the app files
> FROM ${APP_ENV}-preinstall as install
> RUN echo "running install commands"
> 
> FROM install as prod-postinstall
> RUN echo "running PROD post-install commands"
> 
> FROM install as dev-postinstall
> RUN echo "running DEV post-install commands"
> 
> FROM ${APP_ENV}-postinstall as final
> RUN echo "running final commands"
> EOF
> Only the DEV build-stages are executed:
> 
> ```
> [+] Building 1.8s (9/9) FINISHED                                         
>  => local://dockerfile (Dockerfile)                                 0.0s
>  => => transferring dockerfile: 705B                                0.0s
>  => local://context (.dockerignore)                                 0.0s
>  => => transferring context: 02B                                    0.0s
>  => docker-image://docker.io/library/alpine:latest                  0.0s
>  => CACHED /bin/sh -c echo "running BASE commands"                  0.0s
>  => /bin/sh -c echo "running DEV pre-install commands"              0.4s
>  => /bin/sh -c echo "running install commands"                      0.5s
>  => /bin/sh -c echo "running DEV post-install commands"             0.4s
>  => /bin/sh -c echo "running final commands"                        0.4s
>  => exporting to image                                              0.0s
>  => => exporting layers                                             0.0s
>  => => writing image sha256:1cf6caaaab4689f6b11ea4d61b5b6175e54074  0.0s
>  => => naming to docker.io/library/dev                              0.0s
> ```

To this day, this is not clearly documented.
